+++
title = "Chromosome overview plots"

date = 2020-01-27T00:00:00
# lastmod = 2019-04-19T00:00:00

draft = false  # Is this a draft? true/false
toc = true  # Show table of contents? true/false
type = "docs"  # Do not modify.
+++

```{r setup, eval=FALSE, echo=FALSE}
library(tidyverse)
library(ggbio)
library(AnnotationDbi)
library(GenomicRanges)
library(Biostrings)
library(GenomicFeatures)
library(data.table)
```

## Introduction

In this tutorial we will use the [ggbio](https://bioconductor.org/packages/release/bioc/html/ggbio.html) package to build overview plots. 

These are good for showing a simple representation of ChIP-seq and other datasets.

![Example of a simple overview plot. In this plot genes of the positive (red) and negative (blue) strands of a single bacterial chromosome are shown as rectangles on two tracks.](/tutorial/circle_files/example.png)

In the example above only two tracks are shown, which are a representation of the all the genes encoded on the chromosome. Inner tracks can then be added to represent over types of data that have been aligned to the genome, such as sequencing reads.

## Required packages and data

Install (if necessary) and then load the following packages:

- tidyverse  
- data.table  
  
These packages are bioconductor packages:  

- ggbio  
- AnnotationDbi  
- GenomicRanges  
- Biostrings  
- GenomicFeatures  

You will also need to import annotation data for the chromosome(s) as a TxDB. You can use GFF to make a TxDB or download a TxDB that already exists for the organism.

### Create a TxDB object

A TxDB object contains genome annotation information. It is very simple to make, once you have done you can save and reload it whenever you wish to use it. 

- Download a GFF or GTF annotation file (from NCBI)

```{r TxdBcreation, eval=FALSE, echo=TRUE}
library(GenomicFeatures)
library(AnnotationDbi)
<genomeName>TxDB <- makeTxDbFromGFF("annotation<genomeName>.gff")
saveDb(<genomeName>TxDB, file="<genomeName>TxDB.sqlite")
```

- Loading the TxDB file (replacing <genome>, with a name for the genome you are using):

```{r loadtxdb, eval=FALSE, echo=TRUE}
<genome>Annotation <- loadDb("<genomeName>TxDB.sqlite")
```

## Creating a gene track

This track shows the genes on the positive and negative strand of the chromosome. To do this in a single track you need to collapse it, because there are often overlapping genes in different reading frames.  

```{r pos_strand, eval=FALSE, echo =TRUE}
p_strand <- as_tibble(cds(<genome>Annotation, filter = list(tx_strand = "+"))) %>%
  dplyr::select(c('seqnames', 'start', 'end')) %>%
  arrange(start) %>%
  dplyr::group_by(g = cumsum(cummax(lag(end, default = first(end))) < start)) %>% 
  dplyr::summarise(start = first(start), end = max(end)) %>%
  dplyr::select(c('start', 'end'))

p_strand$seqnames <- "NC_000000.0"

n_strand <- as_tibble(cds(chrIAnnotation, filter = list(tx_strand = "-"))) %>%
  dplyr::select(c('seqnames', 'start', 'end')) %>%
  dplyr::arrange(start) %>%
  dplyr::group_by(g = cumsum(cummax(lag(end, default = first(end))) < start)) %>% 
  dplyr::summarise(start = first(start), end = max(end)) %>%
  dplyr::select(c('start', 'end'))

n_strand$seqnames <- "NC_000000.0"
```

First, import the coding sequences from the annotation file and filter for the positive strand only.  

Then arrange data frame by the start position of each gene.   

The group_by and summarise functions are used to 'collapse' the data, by using the start and end positions of genes to merge overlapping coding regions in the different reading frames.  

`lag(end)` adds a column that tracks the previous end values, `cummax` stores the highest of previous end values which is then compared with the end. When a start value is higher than the maximum of the previous stop, it creates a new group. `cumsum` tracks the number of TRUEs and gives each different group (i.e.. overlapping genes form groups) a unique number. `summarise` aggregates based on the group id using the first start value of each group and the largest end value.  

Finally need to add back in the `seqname` column. This should be the RefSeq value from NCBI and should match up the chromosome name used in the reference genome. This is repeated for the negative strand.  

Next you create GRanges objects with the `p_strand` and `n_strand` information using the `makeGRangesFromDataFrame` function from the `GenomicRanges` package.

```{r GRanges, eval=FALSE, echo=TRUE}
pos_genes <- makeGRangesFromDataFrame(p_strand)
neg_genes <- makeGRangesFromDataFrame(n_strand)
```

## Creating the plot

The plot is created using the [ggbio](https://bioconductor.org/packages/release/bioc/html/ggbio.html) package. It works similarly to ggplot, so you can add multiple elements by using '+' and various geoms (such as point, line, link, etc..) For the gene track we will use the rectangle (`rect`) geom.

```{r ggbioplot, eval=FALSE, echo=TRUE}
plot <- ggbio(buffer = 0.2, radius = 10) +
  #circle(avg_reads_track, geom = "line", trackWidth= 3,
  #      color = "dodgerblue2", aes(y=average), buffer=35, radius=6, space.skip=0) +
  circle(neg_genes, geom = "rect", color=NA, 
        fill='blue', trackWidth=0.5, space.skip=0) +
  circle(pos_genes, geom = "rect", color=NA, 
         fill='red', trackWidth=0.5, space.skip=0) 
plot
```

The plot is initialised using a call to the `ggbio` function. The two parameters:  

- buffer: the space between each track  
- radius: the size of the circle plot (might need to increase this depending on the number of tracks you want to add)  

The `circle()` function is called for creating a track in the plot, there a number of parameters that can be set to alter the aesthetics of the track. You can view the help file and all the parameters that can be altered by entering `?circle` in the R console. 

![Help page for the ggbio circle layout function.](/tutorial/circle_files/circle_help.png)

The `space.skip` parameter set to zero means that there is no space between the beginning and end of the chromosome. This is required because the default behaviour is designed to have space between multiple chromosomes. 

```{r ChIP_seq_plots, eval=FALSE, include=FALSE}
# Genome axis
axisTrack <- GenomeAxisTrack(background.title='white', fontfamily="Open Sans")
options(ucscChromosomeNames = FALSE)

#import bed files

peaks.rep1 <- import.bed("~/Google Drive/GraingerLab/VpsT_ChIPseq/macs/macs_chrI_r1_summits.bed")
peaks.rep2 <- import.bed("~/Google Drive/GraingerLab/VpsT_ChIPseq/macs/macs_chrI_r2_summits.bed")
peaks.merged <- import.bed("~/Google Drive/GraingerLab/VpsT_ChIPseq/macs/macs_chrI_r1andr2_summits.bed")

#VpsT tracks (BAMfile)
VpsT_r1_track <-
  AlignmentsTrack(
    "~/Google Drive/GraingerLab/VpsT_ChIPseq/alignments/chrI/vpst_flag_r1_rmdup.bam",
    name = "VpsT rep 1",
    type = "coverage",
    fill = 'orange',
    col = 'orange'
  )

VpsT_r2_track <-
  AlignmentsTrack(
    "~/Google Drive/GraingerLab/VpsT_ChIPseq/alignments/chrI/vpst_flag_r2_rmdup.bam",
    name = "VpsT rep 2",
    type = "coverage",
    fill = 'green',
    col = 'green'
  )

rmdup2 <- DataTrack(range="~/Google Drive/GraingerLab/VpsT_ChIPseq/alignments/chrI/vpst_flag_r2_rmdup.bam", name="rmdup",
                    window=-500, importFunction=strandedBamImport,
                    stream=TRUE,
                    type = "polygon",
                    col.mountain = '#21B7D2',
                    fill.mountain = c('#21B7D2', '#21B7D2'))

rmdup1 <- DataTrack(range="~/Google Drive/GraingerLab/VpsT_ChIPseq/alignments/chrI/vpst_flag_r1_rmdup.bam", name="rmdup",
                    window=-500, importFunction=strandedBamImport,
                    stream=TRUE,
                    type = "polygon",
                    col.mountain = '#21B7D2',
                    fill.mountain = c('#21B7D2', '#21B7D2'))

bam1 <- DataTrack(range="~/Google Drive/GraingerLab/VpsT_ChIPseq/alignments/chrI/vpst_flag_r1_49f22bd15de9.bam", name="rep1",
                    window=-500, importFunction=strandedBamImport,
                    stream=TRUE,
                    type = "polygon",
                    ylim= c(-60,60),
                    col.mountain = '#21B7D2',
                    fill.mountain = c('#21B7D2', '#21B7D2'))

bam2 <- DataTrack(range="~/Google Drive/GraingerLab/VpsT_ChIPseq/alignments/chrI/vpst_flag_r2_49f23e6c8278.bam", name="rep2",
                    window=-500, importFunction=strandedBamImport,
                    stream=TRUE,
                    type = "polygon",
                    ylim= c(-6000,6000),
                    col.mountain = '#21B7D2',
                    fill.mountain = c('#21B7D2', '#21B7D2'))

#Annotation (genes)
annotation_track <- GeneRegionTrack(
  chrIAnnotation,
  name = "Genes",
  shape = "arrow",
  collapseTranscripts = TRUE,
  exonAnnotation = "gene",
  fontcolor = "white",
  size = 4.5,
  col = '#27487C',
  fontcolor.item = "white",
  fontface= 'italic',
  fill = '#27487C',
  fontsize = 9,
  fontfamily = "Open Sans"
)

rmdup_avg_reads_wig <- import("~/Google Drive/GraingerLab/VpsT_ChIPseq/wig_files/chrI/average_reads__rmdup_05_30.wig")

tg_average_reads_wig <- import("~/Google Drive/GraingerLab/VpsT_ChIPseq/wig_files/chrI/15_02average_reads.wig")

## hns data
hns_wig_chr1 <- import("~/Google Drive/GraingerLab/hns/chr1.wig.gz")

hns.track <-
  DataTrack(
    range = hns_wig_chr1,
    name = "H-NS",
    type = "polygon",
    col.mountain = '#1FB524',
    fill.mountain = c('#1FB524', '#1FB524')
  )

rmdup.track <-
  DataTrack(
    range = rmdup_avg_reads_wig,
    name = "VpsT (dups removed)",
    window = 100,
    ylim=c(0,2),
    type = "polygon",
    col.mountain = '#21B7D2',
    fill.mountain = c('#21B7D2', '#21B7D2')
  )


counts.track <-
  DataTrack(
    range = tg_average_reads_wig,
    name = "VpsT",
    window = 500,
    ylim= c(0,50),
    type = "polygon",
    col.mountain = '#21B7D2',
    fill.mountain = c('#21B7D2', '#21B7D2')
  )


#Plots
998899	999240
# use the makeplot function to produce gviz plots for specified genomic regions
# provide a name, and the from and to co-ordinates
# plots have axisTrack, annotation_track and counts.track, for other tracks build
# manually
makeplot <- function(name, from, to) {
  name <- plotTracks(
    list(annotation_track, bam1, bam2),
    from = from, 
    to = to, 
    chromosome = "NC_002505.1",
    background.title = "white",
    col.title = 'black',
    col.axis = "black",
    fontfamily.title = "Open Sans SemiBold"
  )
}


# makeplot code can be used for exporting to powerpoint files, see below

makeplot(VC0364, from = 382048, to = 383948)
makeplot(operon_510, from = 539085, to = 550150)
makeplot(VC0534, from = 563000, to = 565171)
makeplot(VC0825, from = 887472, to = 890472)
makeplot(VC0829, from = 890494, to = 892494)
makeplot(VC0841, from = 901429, to = 905429)
makeplot(VC0917, from = 978211, to = 980211)
makeplot(VC0928, from = 989914, to = 991914)
makeplot(VC0932, from = 998012, to = 999912)
makeplot(VC1029, from = 1095963, to = 1097963)
makeplot(VC1329, from = 1410874, to = 1414874)
makeplot(frh_operon, from = 1731800, to = 1745040)
makeplot(VC1710, from = 1843392, to = 1847392)
makeplot(VC1802, from = 1946982, to = 1948982)
makeplot(VC1851, from = 1989004, to = 1991004)
makeplot(VC2065, from = 2220276, to = 2222276)
makeplot(VC2188, from = 2335166, to = 2337166)
makeplot(VC2222, from = 2375037, to = 2377037)
makeplot(VC2667, from = 2834735, to = 2836735)




makeplot(vpv_operon, from = 2637368, to = 2641100)
makeplot(frh_operon, from = 1731800, to = 1745040)
makeplot(operon_510, from = 539085, to = 550150)
makeplot(rpos, from = 998000, to = 1000700)
makeplot(target_1304, from = 1383845, to = 1387248)


overview <- plotTracks(
  list(axisTrack, counts.track, peaks1.track, peaks2.track, peaksmerged.track),
  from = 1,
  to = 2961149,
  showFeatureId = TRUE,
  collapseTranscripts = TRUE,
  background.title = "white",
  col.title = 'black',
  col.axis = "black",
  fontfamily.title = "Open Sans SemiBold"
)

VC510_VC512 <- plotTracks(
  list(axisTrack, annotation_track, VpsT_r1_track, counts.track, dg.counts.track, peaks1.track, peaks2.track, peaksmerged.track, bdg.rep1.track, bdg.rep2.track),
  from = 539085,
  to = 550150,
  chromosome = "NC_002505.1",
  background.title = "white",
  col.title = 'black',
  col.axis = "black",
  fontfamily.title = "Open Sans SemiBold"
)


#ChIPseeker
#annotate peaks uses BED files to find the nearest gene
rep1.annopeak <- annotatePeak(peaks.rep1, TxDb = chrIAnnotation)
df.rep1.annopeak <- as.data.frame(rep1.annopeak)
write_delim(df.rep1.annopeak, path = "~/Google Drive/GraingerLab/VpsT_ChIPseq/annotation/ChIPseeker_rep1.txt")

rep2.annopeak <- annotatePeak(peaks.rep2, TxDb = chrIAnnotation)
df.rep2.annopeak <- as.data.frame(rep2.annopeak)
write_delim(df.rep2.annopeak, path = "~/Google Drive/GraingerLab/VpsT_ChIPseq/annotation/ChIPseeker_rep2.txt")

merged.annopeak <- annotatePeak(peaks.merged, TxDb = chrIAnnotation)
df.merged.annopeak <- as.data.frame(merged.annopeak)
write_delim(df.merged.annopeak, path = "~/Google Drive/GraingerLab/VpsT_ChIPseq/annotation/ChIPseeker_merged.txt")

plotDistToTSS(merged.annopeak)

#histograms showing peak distance from nearest TSS
qplot(df.merged.annopeak$distanceToTSS, binwidth=50)
qplot(df.rep2.annopeak$distanceToTSS, binwidth=50)
qplot(df.rep1.annopeak$distanceToTSS, binwidth=50)

# export gviz plots to powerpoint file in diagrams folder
# reads in the gvizplots.pptx file
# add new slide with add_slide
# then add plot to the new slide with ph_with_vg function
doc <- read_pptx("~/Google Drive/GraingerLab/diagrams/gvizplots.pptx")
doc <- add_slide(doc, layout = "Title and Content", master = "Office Theme")
doc <- ph_with_vg(doc, code=
                    plotTracks(
    list(axisTrack, annotation_track, counts.track),
    from = 1383845, 
    to = 1387248, 
    chromosome = "NC_002505.1",
    background.title = "white",
    col.title = 'black',
    col.axis = "black",
    fontfamily.title = "Open Sans SemiBold"
  ),                  type="body")
  print(doc, target = "~/Google Drive/GraingerLab/diagrams/gvizplots.pptx")

#doc <- add_slide(doc, layout = "Title and Content", master = "Office Theme")
#doc <- ph_with_vg(doc, code=
#makeplot(frh_operon, from = 1731800,to = 1745040), type = "body")
```



