<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 4.1.0">
  <meta name="generator" content="Hugo 0.55.6" />

  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Tom Guest">

  
  
  
    
  
  <meta name="description" content="Introduction In this tutorial we will use the QuasR package to align our sequencing data to a reference genome.
QuasR uses the alignment program Bowtie, and will produce bam files for each alignment. There are more advanced options for including auxiliary genomes (which are used to align &lsquo;leftover&rsquo; unmapped sequences) which are useful to check for contaminating DNA and for spiked experiments.
Sequencing reads In ChIP-seq the sequencing reads are typically short single end dsDNA.">

  
  <link rel="alternate" hreflang="en-us" href="https://tomguest.netlify.com/tutorial/alignment/">

  


  

  

  

  

  

  

  
  
  
  <meta name="theme-color" content="#f2671d">
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.0/css/all.css" integrity="sha384-aOkxzJ5uQz7WBObEZcHvV5JvRW3TUc2rNPA7pe3AwnsUohiw1Vj2Rgx2KSOkF5+h" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" crossorigin="anonymous" title="hl-light">
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/dracula.min.css" crossorigin="anonymous" title="hl-dark" disabled>
        
      
    

    

    

  

  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Open+Sans|Pontano+Sans|Roboto+Mono">
  

  <link rel="stylesheet" href="/styles.css">
  

  
  
  

  
  <link rel="alternate" href="https://tomguest.netlify.com/index.xml" type="application/rss+xml" title="Tom Guest ">
  <link rel="feed" href="https://tomguest.netlify.com/index.xml" type="application/rss+xml" title="Tom Guest ">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="https://tomguest.netlify.com/tutorial/alignment/">

  
  
  
  
    
    
  
  <meta property="twitter:card" content="summary">
  
  <meta property="twitter:site" content="@tom_guest">
  <meta property="twitter:creator" content="@tom_guest">
  
  <meta property="og:site_name" content="Tom Guest ">
  <meta property="og:url" content="https://tomguest.netlify.com/tutorial/alignment/">
  <meta property="og:title" content="Aligning to a reference genome | Tom Guest ">
  <meta property="og:description" content="Introduction In this tutorial we will use the QuasR package to align our sequencing data to a reference genome.
QuasR uses the alignment program Bowtie, and will produce bam files for each alignment. There are more advanced options for including auxiliary genomes (which are used to align &lsquo;leftover&rsquo; unmapped sequences) which are useful to check for contaminating DNA and for spiked experiments.
Sequencing reads In ChIP-seq the sequencing reads are typically short single end dsDNA."><meta property="og:image" content="https://tomguest.netlify.com/img/icon-192.png">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2019-04-19T00:00:00&#43;01:00">
  
  <meta property="article:modified_time" content="2019-04-19T00:00:00&#43;01:00">
  

  

  

  <title>Aligning to a reference genome | Tom Guest </title>

</head>
<body id="top" data-spy="scroll" data-target="#TableOfContents" data-offset="71" >
  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" role="textbox" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


<nav class="navbar navbar-light fixed-top navbar-expand-lg py-0" id="navbar-main">
  <div class="container">

    
      <a class="navbar-brand" href="/">Tom Guest </a>
      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
        <span><i class="fas fa-bars"></i></span>
      </button>
      

    
    <div class="collapse navbar-collapse" id="navbar">

      
      
      <ul class="navbar-nav mr-auto">
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#projects">
            
            <span>Learning R</span>
            
          </a>
        </li>

        
        

      
      </ul>
      <ul class="navbar-nav ml-auto">
      

        

        
        <li class="nav-item">
          <a class="nav-link js-search" href="#"><i class="fas fa-search" aria-hidden="true"></i></a>
        </li>
        

        

        
        <li class="nav-item">
          <a class="nav-link js-dark-toggle" href="#"><i class="fas fa-moon" aria-hidden="true"></i></a>
        </li>
        

      </ul>

    </div>
  </div>
</nav>



<div class="container-fluid docs">
  <div class="row flex-xl-nowrap">
    <div class="col-12 col-md-3 col-xl-2 docs-sidebar">
      




<form class="docs-search d-flex align-items-center">
  <button class="btn docs-toggle d-md-none p-0 mr-3" type="button" data-toggle="collapse" data-target="#docs-nav" aria-controls="docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    <span><i class="fas fa-bars"></i></span>
  </button>

  
  <input name="q" type="search" class="form-control" id="search-query" placeholder="Search..." autocomplete="off">
  
</form>

<nav class="collapse docs-links" id="docs-nav">
  
  
  <div class="docs-toc-item">
    <a class="docs-toc-link" href="/tutorial/">ChIP-seq with R</a>
    <ul class="nav docs-sidenav">
      
      <li >
        <a href="/tutorial/">Overview</a>
      </li>
      
      <li >
        <a href="/tutorial/bioconductor_intro/">Introduction to Bioconductor</a>
      </li>
      
      <li >
        <a href="/tutorial/bsgenome/">Forging a BSgenome package</a>
      </li>
      
      <li class="active">
        <a href="/tutorial/alignment/">Aligning to a reference genome</a>
      </li>
      
      <li >
        <a href="/tutorial/normalisation/">Normalising read counts</a>
      </li>
      
      <li >
        <a href="/tutorial/peakcalling/">Peak calling</a>
      </li>
      
      <li >
        <a href="/tutorial/motif/">De-novo motif finding</a>
      </li>
      
    </ul>
    

  </div>
  
  
</nav>

    </div>

    
    <div class="d-none d-xl-block col-xl-2 docs-toc">
      
      <p class="docs-toc-title">On this page</p>
      

      <nav id="TableOfContents">
<ul>
<li>
<ul>
<li><a href="#introduction">Introduction</a>
<ul>
<li><a href="#sequencing-reads">Sequencing reads</a></li>
</ul></li>
<li><a href="#install-the-quasr-package">Install the QuasR package</a></li>
<li><a href="#folder-structure">Folder structure</a></li>
<li><a href="#preparation">Preparation</a></li>
<li><a href="#making-the-alignment-s">Making the alignment(s)</a>
<ul>
<li><a href="#auxiliary-alignments">Auxiliary alignments</a></li>
</ul></li>
<li><a href="#alignment-stats-and-quality-reports">Alignment stats and quality reports</a></li>
</ul></li>
</ul>
</nav>

      <ul class="nav toc-top">
        <li><a href="#">Back to top</a></li>
      </ul>

      
    </div>
    

    <main class="col-12 col-md-9 col-xl-8 py-md-3 pl-md-5 docs-content" role="main">

      <article class="article" itemscope itemtype="http://schema.org/Article">

        <div class="docs-article-container">
          <h1 itemprop="name">Aligning to a reference genome</h1>

          <div class="article-style" itemprop="articleBody">
            

<h2 id="introduction">Introduction</h2>

<p>In this tutorial we will use the <a href="https://bioconductor.org/packages/release/bioc/html/QuasR.html" target="_blank">QuasR</a> package to align our sequencing data to a reference genome.</p>

<p>QuasR uses the alignment program <code>Bowtie</code>, and will produce <code>bam</code> files for each alignment. There are more advanced options for including auxiliary genomes (which are used to align &lsquo;leftover&rsquo; unmapped sequences) which are useful to check for contaminating DNA and for spiked experiments.</p>

<h3 id="sequencing-reads">Sequencing reads</h3>

<p>In ChIP-seq the sequencing reads are typically short single end dsDNA. This means that the 5’ end will be sequenced on “+” strand and the 3’ end will be on “-” strand. &ldquo;+” reads extend only in positive direction and “-” reads in negative direction which results in the typical bimodal peak at transcription factor binding sites.</p>

<p><img src="/tutorial/alignment_files/reads.png" alt="" /></p>

<h2 id="install-the-quasr-package">Install the QuasR package</h2>

<pre><code class="language-r">if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;QuasR&quot;, version = &quot;3.8&quot;)
</code></pre>

<h2 id="folder-structure">Folder structure</h2>

<p>I recommend having a folder for each new project. Within this create separate folders for <code>raw_fastq</code>, <code>alignments</code>, <code>genomes</code>, etc. You can then keep things tidy and avoid making any changes or accidentally deleting your raw fastq sequencing data.</p>

<p>You should set the project root as working directory, (or even better create a RStudio project).</p>

<pre><code class="language-r">setwd(&quot;path/to/myChIPseqProject&quot;)
</code></pre>

<h2 id="preparation">Preparation</h2>

<p>Download and save the genome <code>fasta</code> file(s) to your genomes folder. Then direct R to their location(s).</p>

<pre><code class="language-r">chromosomeI &lt;- &quot;genomes/chromosome_I.fasta&quot;
chromosomeII &lt;- &quot;genomes/chromosome_II.fasta&quot;
</code></pre>

<p>You will also need to create a tabulated <code>sampleFile.txt</code> which should contain two columns <code>FileName</code> and <code>SampleName</code>, list the names of the files (exactly as they are) and a name for each sample.</p>

<p>If you multiple replicates you can either name them as <code>replicate_**</code> or if you give them all the same SampleName, R will know to treat them as replicates, but you will need to decide if that is how you want R to behave. If you are unsure stick with the former approach.</p>

<table>
<thead>
<tr>
<th>FileName</th>
<th>SampleName</th>
</tr>
</thead>

<tbody>
<tr>
<td>sample1.fastq</td>
<td>sample1</td>
</tr>

<tr>
<td>sample2.fastq</td>
<td>sample2</td>
</tr>
</tbody>
</table>

<p>Then assign this to <code>sampleFile</code> in R:</p>

<pre><code class="language-r">sampleFile &lt;- &quot;raw_fastq/sampleFile.txt&quot;
</code></pre>

<h2 id="making-the-alignment-s">Making the alignment(s)</h2>

<p>If you have multiple chromosomes you can assign the alignment to each as a separate QuasR project.</p>

<div class="alert alert-note">
  <div>
    
You could choose to align using a BSgenome object instead, see the [QuasR documentation](https://bioconductor.org/packages/release/bioc/vignettes/QuasR/inst/doc/QuasR.html#617_using_a_bsgenome_package_as_reference_genome) if you prefer to do that.

  </div>
</div>

<p>Before you run the <code>qAlign()</code> function, make sure that you have created the folders to store your alignments (or you will get an error saying the directory does not exist).</p>

<p>To see a full list of other alignment parameters, use <code>?qAlign</code> to see the helpfile.</p>

<pre><code class="language-r">proj_chrI &lt;- qAlign(sampleFile, genome=chromosomeI, projectName = &quot;chrI&quot;, alignmentsDir= &quot;alignments/chrI/&quot;)
proj_chrII &lt;- qAlign(sampleFile, genome=chromosomeII, projectName = &quot;chrII&quot;, alignmentsDir= &quot;alignments/chrII/&quot;)
</code></pre>

<p>If you run the same code again and you already have the output <code>.bam</code> and <code>.txt</code> files it will not repeat the alignment. If you alter any alignment parameters it will do a new alignment.</p>

<p>You can type the project name into the console to see which output files correspond to each alignment.</p>

<pre><code class="language-r">proj_chrI
proj_chrII
</code></pre>

<h3 id="auxiliary-alignments">Auxiliary alignments</h3>

<p>You can specify auxiliary genomes to be used to align unmapped sequences from the core genome. For example, if you have a spiked experiment or to check for contaminating sequences.</p>

<p>To do this supply a text file with a list of additional genome files like this:</p>

<table>
<thead>
<tr>
<th>FileName</th>
<th>AuxName</th>
</tr>
</thead>

<tbody>
<tr>
<td>NC_001422.1.fa</td>
<td>phiX174</td>
</tr>
</tbody>
</table>

<p>Then assign it in R:</p>

<pre><code class="language-r">additionalGenomes &lt;- &quot;genomes/additionalGenomes.txt&quot;
</code></pre>

<p>And then when running qAlign, add <code>auxiliaryFile = additionalGenomes</code> to the arguments list.</p>

<pre><code class="language-r">proj_chrI &lt;- qAlign(sampleFile, genome=chromosomeI, projectName = &quot;chrI&quot;, alignmentsDir= &quot;alignments/chrI/&quot;,
                    auxiliaryFile = additionalGenomes)
</code></pre>

<h2 id="alignment-stats-and-quality-reports">Alignment stats and quality reports</h2>

<p>You can use the <code>alignmentStats()</code> function to find out the number of mapped/unmapped reads.</p>

<pre><code class="language-r">alignmentStats(proj_chrI)
alignmentStats(proj_chrII)
</code></pre>

<p>You can produce PDF quality reports for each alignment using the <code>qQCReport()</code> function.</p>

<pre><code class="language-r">qQCReport(proj_chrI, pdfFilename = &quot;quality_reports/chrI_quality.pdf&quot;)
qQCReport(proj_chrII, pdfFilename = &quot;quality_reports/chrII_quality.pdf&quot;)
</code></pre>

          </div>

          

        </div>

        <div class="body-footer">
          Last updated on Apr 19, 2019
        </div>

      </article>

      <footer class="site-footer">
  

  <p class="powered-by">
    

    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
  </p>
</footer>


    </main>
  </div>
</div>

    

    
    

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha512-+NqPlbbtM1QqiK8ZAo4Yrj2c4lNQoGv8P79DPtKzj++l5jnN39rHA/xsqn8zE9l0uSoxaCdrOgFs6yjyfbBxSg==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha256-VsEqElsCHSGmnmHXGQzvoWjWwoznFSZc6hs7ARLRacQ=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
        
        <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
        
      

      
      
    

    
    

    
    
    

    
    
    <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script>
      const search_index_filename = "/index.json";
      const i18n = {
        'placeholder': "Search...",
        'results': "results found",
        'no_results': "No results found"
      };
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/4.1.1/anchor.min.js" integrity="sha256-pB/deHc9CGfFpJRjC43imB29Rse8tak+5eXqntO94ck=" crossorigin="anonymous"></script>
    <script>
      anchors.add();
    </script>
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/js/academic.min.d7381f2d79e6271d4da28f474f49096c.js"></script>

  </body>
</html>


