<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 4.1.0">
  <meta name="generator" content="Hugo 0.55.6" />

  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Tom Guest">

  
  
  
    
  
  <meta name="description" content="Introduction In this tutorial we will use the BSgenome package to create an R package that contains the genome for Vibrio cholerae, but you can replace it with your favourite organism.
The BSgenome package provides a framework for interacting with genome information in R.
 There may already be a BSgenome package for your favourite organism. Check the [list](https://kasperdanielhansen.github.io/genbioconductor/html/BSgenome.html) of available genomes first.   Once you have forged (created) and installed the package, you will be able to load the genome as you would any other R package, like this:">

  
  <link rel="alternate" hreflang="en-us" href="https://tomguest.netlify.com/tutorial/bsgenome/">

  


  

  

  

  

  

  

  
  
  
  <meta name="theme-color" content="#f2671d">
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.0/css/all.css" integrity="sha384-aOkxzJ5uQz7WBObEZcHvV5JvRW3TUc2rNPA7pe3AwnsUohiw1Vj2Rgx2KSOkF5+h" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" crossorigin="anonymous" title="hl-light">
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/dracula.min.css" crossorigin="anonymous" title="hl-dark" disabled>
        
      
    

    

    

  

  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Open+Sans|Pontano+Sans|Roboto+Mono">
  

  <link rel="stylesheet" href="/styles.css">
  

  
  
  

  
  <link rel="alternate" href="https://tomguest.netlify.com/index.xml" type="application/rss+xml" title="Tom Guest ">
  <link rel="feed" href="https://tomguest.netlify.com/index.xml" type="application/rss+xml" title="Tom Guest ">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="https://tomguest.netlify.com/tutorial/bsgenome/">

  
  
  
  
    
    
  
  <meta property="twitter:card" content="summary">
  
  <meta property="twitter:site" content="@tom_guest">
  <meta property="twitter:creator" content="@tom_guest">
  
  <meta property="og:site_name" content="Tom Guest ">
  <meta property="og:url" content="https://tomguest.netlify.com/tutorial/bsgenome/">
  <meta property="og:title" content="Forging a BSgenome package | Tom Guest ">
  <meta property="og:description" content="Introduction In this tutorial we will use the BSgenome package to create an R package that contains the genome for Vibrio cholerae, but you can replace it with your favourite organism.
The BSgenome package provides a framework for interacting with genome information in R.
 There may already be a BSgenome package for your favourite organism. Check the [list](https://kasperdanielhansen.github.io/genbioconductor/html/BSgenome.html) of available genomes first.   Once you have forged (created) and installed the package, you will be able to load the genome as you would any other R package, like this:"><meta property="og:image" content="https://tomguest.netlify.com/img/icon-192.png">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2018-09-09T00:00:00&#43;01:00">
  
  <meta property="article:modified_time" content="2018-09-09T00:00:00&#43;01:00">
  

  

  

  <title>Forging a BSgenome package | Tom Guest </title>

</head>
<body id="top" data-spy="scroll" data-target="#TableOfContents" data-offset="71" >
  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" role="textbox" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


<nav class="navbar navbar-light fixed-top navbar-expand-lg py-0" id="navbar-main">
  <div class="container">

    
      <a class="navbar-brand" href="/">Tom Guest </a>
      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
        <span><i class="fas fa-bars"></i></span>
      </button>
      

    
    <div class="collapse navbar-collapse" id="navbar">

      
      
      <ul class="navbar-nav mr-auto">
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#projects">
            
            <span>Learning R</span>
            
          </a>
        </li>

        
        

      
      </ul>
      <ul class="navbar-nav ml-auto">
      

        

        
        <li class="nav-item">
          <a class="nav-link js-search" href="#"><i class="fas fa-search" aria-hidden="true"></i></a>
        </li>
        

        

        
        <li class="nav-item">
          <a class="nav-link js-dark-toggle" href="#"><i class="fas fa-moon" aria-hidden="true"></i></a>
        </li>
        

      </ul>

    </div>
  </div>
</nav>



<div class="container-fluid docs">
  <div class="row flex-xl-nowrap">
    <div class="col-12 col-md-3 col-xl-2 docs-sidebar">
      




<form class="docs-search d-flex align-items-center">
  <button class="btn docs-toggle d-md-none p-0 mr-3" type="button" data-toggle="collapse" data-target="#docs-nav" aria-controls="docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    <span><i class="fas fa-bars"></i></span>
  </button>

  
  <input name="q" type="search" class="form-control" id="search-query" placeholder="Search..." autocomplete="off">
  
</form>

<nav class="collapse docs-links" id="docs-nav">
  
  
  <div class="docs-toc-item">
    <a class="docs-toc-link" href="/tutorial/">ChIP-seq with R</a>
    <ul class="nav docs-sidenav">
      
      <li >
        <a href="/tutorial/">Overview</a>
      </li>
      
      <li >
        <a href="/tutorial/bioconductor_intro/">Introduction to Bioconductor</a>
      </li>
      
      <li class="active">
        <a href="/tutorial/bsgenome/">Forging a BSgenome package</a>
      </li>
      
      <li >
        <a href="/tutorial/alignment/">Aligning to a reference genome</a>
      </li>
      
      <li >
        <a href="/tutorial/normalisation/">Normalising read counts</a>
      </li>
      
      <li >
        <a href="/tutorial/peakcalling/">Peak calling</a>
      </li>
      
      <li >
        <a href="/tutorial/motif/">De-novo motif finding</a>
      </li>
      
    </ul>
    

  </div>
  
  
</nav>

    </div>

    
    <div class="d-none d-xl-block col-xl-2 docs-toc">
      
      <p class="docs-toc-title">On this page</p>
      

      <nav id="TableOfContents">
<ul>
<li>
<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#install-the-bsgenome-package">Install the BSgenome package</a></li>
<li><a href="#download-the-genome-sequence">Download the genome sequence</a></li>
<li><a href="#prepare-your-files">Prepare your files</a>
<ul>
<li><a href="#seed-file">Seed file</a></li>
</ul></li>
<li><a href="#forge-the-package">Forge the package</a></li>
<li><a href="#install-your-genome-package">Install your genome package</a></li>
<li><a href="#loading-and-accessing-the-genome-in-r">Loading and accessing the genome in R</a></li>
</ul></li>
</ul>
</nav>

      <ul class="nav toc-top">
        <li><a href="#">Back to top</a></li>
      </ul>

      
    </div>
    

    <main class="col-12 col-md-9 col-xl-8 py-md-3 pl-md-5 docs-content" role="main">

      <article class="article" itemscope itemtype="http://schema.org/Article">

        <div class="docs-article-container">
          <h1 itemprop="name">Forging a BSgenome package</h1>

          <div class="article-style" itemprop="articleBody">
            

<h2 id="introduction">Introduction</h2>

<p>In this tutorial we will use the BSgenome package to create an R package that contains the genome for <em>Vibrio cholerae</em>, but you can replace it with <strong>your favourite organism</strong>.</p>

<p>The <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html" target="_blank">BSgenome</a> package provides a framework for interacting with genome information in R.</p>

<div class="alert alert-warning">
  <div>
    
There may already be a BSgenome package for your favourite organism. Check the [list](https://kasperdanielhansen.github.io/genbioconductor/html/BSgenome.html) of available genomes first.

  </div>
</div>

<p>Once you have <em>forged</em> (created) and installed the package, you will be able to load the genome as you would any other R package, like this:</p>

<pre><code class="language-r">library(BSgenome.Vcholerae.NCBI.N16961)
</code></pre>

<p>If <em>V. cholerae</em> happens to be <strong>your favourite organism</strong> too, and want to save yourself some time, you can <a href="/tutorial/files/BSgenome.Vcholerae.NCBI.N16961_1.0.0.tar.gz">download</a> the package I created, and skip to the &lsquo;Install your genome package&rsquo; section to get started.</p>

<h2 id="install-the-bsgenome-package">Install the BSgenome package</h2>

<pre><code class="language-r">if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;BSgenome&quot;, version = &quot;3.8&quot;)
</code></pre>

<h2 id="download-the-genome-sequence">Download the genome sequence</h2>

<p>You will need to download the fasta files for the genome you want to forge. Since <em>V. cholerae</em> has two chromosomes, these are downloaded separately. You can use whatever source you like (EMBO, NCBI etc..) but make sure you get it in <strong>fasta</strong> file type.</p>

<p>For the <em>V.cholerae</em> reference genome (N16961):</p>

<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_002505.1" target="_blank">chromosome I</a><br /></li>
<li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_002506.1" target="_blank">chromesome II</a><br /></li>
</ul>

<p>At the NCBI website, click send to file, then select FASTA.</p>

<h2 id="prepare-your-files">Prepare your files</h2>

<p>I created a new folder on my Desktop and set this as my working directory:</p>

<pre><code class="language-r">setwd(&quot;~/Desktop/genomepackage&quot;)
</code></pre>

<p>Within this create another named <code>seqs_srcdir</code>, move the fasta files here.</p>

<p>Make sure the files are appropriately named (i.e.. as-is from source) for <em>V. cholerae</em> the two chromosome fasta files are named: <code>NC_002505.1.fa</code> and  <code>NC_002506.1.fa</code>. It is tempting to name them something more readable like &ldquo;chromosome 1&rdquo; but this can cause problems later.</p>

<p>Ensure the file extensions are <code>.fa</code> if they are not already, on Mac double check by right-clicking and choosing &lsquo;Get Info&rsquo; because it could still be <code>.fasta</code> - if so, amend it.</p>

<h3 id="seed-file">Seed file</h3>

<p>The seed file contains all the relevant metadata for the BSgenome package, so it is worth supplying as much information as you can. The easiest way to make a seed file is to edit one that already exists, so you can download my <a href="/tutorial/files/BSgenome.Vcholerae.NCBI.N16961-seed">seed file</a> and use it as a template.</p>

<p>You will need to use a text editor such as TextEdit on Mac (right-click and select open with &gt; TextEdit) or RStudio. Use NCBI to populate the relevant information:</p>

<blockquote>
<p>Package: BSgenome.Vcholerae.NCBI.N16961<br />
Title: Full genome sequence for Vibrio cholerae O1 biovar El Tor str N16961<br />
Description: Full genome sequence for the two chromosomes of Vibrio cholerae El Tor N16961 provided by NCBI<br />
Version: 1.0.0<br />
organism: Vibrio cholerae<br />
common_name: V. cholerae<br />
provider: NCBI<br />
provider_version: ASM674v1<br />
release_date: <sup>2014</sup>&frasl;<sub>02</sub><br />
release_name: N16961<br />
source_url: <a href="https://www.ncbi.nlm.nih.gov/genome/?term=Vibrio%20cholerae" target="_blank">https://www.ncbi.nlm.nih.gov/genome/?term=Vibrio%20cholerae</a><br />
organism_biocview: Vibrio_cholerae<br />
BSgenomeObjname: Vcholerae<br />
seqnames: c(&ldquo;NC_002505.1&rdquo;,&ldquo;NC_002506.1&rdquo;)<br />
seqs_srcdir: /User/Desktop/genomepackage/seqs_srcdir</p>
</blockquote>

<p>For genomes with multiple chromosomes, list them as a vector (see above example). Otherwise <code>seqnames: chromosomenameFileName</code> is sufficient (you can remove the c()).</p>

<p>The <code>BSgenomeObjname</code> is important because this is the name you will use to access the package in R once it has been installed.</p>

<p>Save it as is, and then edit the file name. Be careful with the file extensions, double check using &lsquo;Get Info&rsquo; to ensure it has not been changed to <code>.txt</code> or anything else.</p>

<h2 id="forge-the-package">Forge the package</h2>

<p>The package is forged using the <code>forgeBSgenomeDataPkg</code> function.</p>

<p>Simply use the name of the seed file as the only argument and it will create your package files to the same directory.</p>

<div class="alert alert-warning">
  <div>
    
Double check the sequence files are `.fa` file types and that the details in the seed are correct before running.

  </div>
</div>

<pre><code class="language-r">forgeBSgenomeDataPkg(&quot;BSgenome.Vcholerae.NCBI.N16961-seed&quot;)
</code></pre>

<p>If you need to run the function again, delete the previous package files first.</p>

<h2 id="install-your-genome-package">Install your genome package</h2>

<p>To install the genome package you will need to use the Mac command line (Terminal).</p>

<ul>
<li>Close R</li>
<li>Open Terminal</li>
<li>In Terminal navigate to your working directory:

<ul>
<li>use <code>ls</code> to see list of files in the current directory</li>
<li>use <code>cd</code> to move to a directory (i.e.. <code>cd Desktop</code>)</li>
</ul></li>
<li>Run <code>R CMD build BSgenome.Vcholerae.EBI.N16961</code> to compile the package</li>
<li>Run <code>R CMD check BSgenome.Vcholerae.EBI.N16961.tar.gz</code> to check it</li>
<li>Run <code>R CMD INSTALL BSgenome.Vcholerae.NCBI.N16961_1.0.0.tar.gz</code></li>
</ul>

<div class="alert alert-note">
  <div>
    
If you have downloaded my _V. cholerae N16961_ [package](/tutorial/files/BSgenome.Vcholerae.NCBI.N16961_1.0.0.tar.gz) you will need to navigate to wherever you have saved the file.  Then run:  
`R CMD INSTALL BSgenome.Vcholerae.NCBI.N16961_1.0.0.tar.gz`

  </div>
</div>

<p>And you&rsquo;re done! It should now be ready to use.</p>

<h2 id="loading-and-accessing-the-genome-in-r">Loading and accessing the genome in R</h2>

<p>To use the genome in R you will need to load the package using the <code>library()</code> function.</p>

<pre><code class="language-r">library(BSgenome.Vcholerae.NCBI.N16961)
</code></pre>

<p>Enter the <code>BSgenomeObjname</code> (in this case Vcholerae) to print some general information about the genome to the console.</p>

<pre><code class="language-r">Vcholerae
</code></pre>

<pre><code>## V. cholerae genome:
## # organism: Vibrio cholerae (V. cholerae)
## # provider: NCBI
## # provider version: ASM674v1
## # release date: 2014/02
## # release name: N16961
## # 2 sequences:
## #   NC_002505.1 NC_002506.1                                                
## # (use 'seqnames()' to see all the sequence names, use the '$' or '[[' operator
## # to access a given sequence)
</code></pre>

<p><code>length()</code> tells you how many chromosomes there are</p>

<pre><code class="language-r">length(Vcholerae)
</code></pre>

<pre><code>## [1] 2
</code></pre>

<p><code>Vcholerae$NC_002505.1</code> tells you the length of chromosome I and a bit of its sequence</p>

<pre><code class="language-r">Vcholerae$NC_002505.1
</code></pre>

<pre><code>##   2961149-letter &quot;DNAString&quot; instance
## seq: AGGGTCATTAAATATATATAAAGATCTATATAGAGA...GGCTAGAAAATCGCTTTCCTGTTTTTTCGATCAAGG
</code></pre>

<p><code>alphabetFrequency(Vcholerae$NC_002505.1)</code> shows you the ACGT content of chromosome I</p>

<pre><code class="language-r">alphabetFrequency(Vcholerae$NC_002505.1)
</code></pre>

<pre><code>##      A      C      G      T      M      R      W      S      Y      K      V 
## 769234 703384 708931 779567      0      0      0      0      0      0      0 
##      H      D      B      N      -      +      . 
##      0      0      0     33      0      0      0
</code></pre>

<p>You can extract sequence from specific co-ordinates, for example to select the sequence from position 45 to 65 on chromosome I:</p>

<pre><code class="language-r">Vcholerae$NC_002505.1[45:65]
</code></pre>

<pre><code>##   21-letter &quot;DNAString&quot; instance
## seq: TTAGATCTACTATTAAGGAGC
</code></pre>

<p>You could then store this in a data frame, export the data frame as fasta file (with multiple sequence etc..) and use it somewhere else&hellip;</p>

          </div>

          

        </div>

        <div class="body-footer">
          Last updated on Sep 9, 2018
        </div>

      </article>

      <footer class="site-footer">
  

  <p class="powered-by">
    

    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
  </p>
</footer>


    </main>
  </div>
</div>

    

    
    

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha512-+NqPlbbtM1QqiK8ZAo4Yrj2c4lNQoGv8P79DPtKzj++l5jnN39rHA/xsqn8zE9l0uSoxaCdrOgFs6yjyfbBxSg==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha256-VsEqElsCHSGmnmHXGQzvoWjWwoznFSZc6hs7ARLRacQ=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
        
        <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
        
      

      
      
    

    
    

    
    
    

    
    
    <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script>
      const search_index_filename = "/index.json";
      const i18n = {
        'placeholder': "Search...",
        'results': "results found",
        'no_results': "No results found"
      };
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/4.1.1/anchor.min.js" integrity="sha256-pB/deHc9CGfFpJRjC43imB29Rse8tak+5eXqntO94ck=" crossorigin="anonymous"></script>
    <script>
      anchors.add();
    </script>
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/js/academic.min.d7381f2d79e6271d4da28f474f49096c.js"></script>

  </body>
</html>


